File: Jongwon/노란불신호등/노란불신호등.java
/*
차량 신호등 n개(초->노->빨 순으로 반복)
처음에는 초록불
각 신호의 지속 시간은 신호등마다 다릅니다
모든 신호가 노란불이면 정전
모든 신호등이 노란불이 되는 가장 빠른시간을 리턴
그런 경우가 없으면 -1리턴
초록불 다음이 노란불이니까
*/
import java.util.*;

public class 노란불신호등 {
    public int solution(int[][] signals) {
        int answer = -1;
        int n = signals.length;
        // t는 흘러가는 실제 시간이고 1초부터 차례대로 돌려봄
        for(int t=1; t<3200000; t++){
            boolean yellow = true;

            // 현재 시간 t에 모든 신호등이 노란불인지 하나씩 검사함
            for(int i=0; i<n; i++){
                int g = signals[i][0];
                int y = signals[i][1];
                int r = signals[i][2];

                int c = g+y+r;
                int rr = t%c;
                if (rr == 0) rr = c;

                // 하나라도 노란불 구간이 아니라면 탈락 처리 후 다음 시간으로
                if(!(rr > g && rr<=g+y)){
                    yellow = false;
                    break;
                }
            }

            if(yellow){
                return t;
            }
        }


        return answer;
    }
}

File: Suho/노란불신호등/노란불신호등.py
from math import gcd

def solution(signals):
    answer = 0
    #1열 : G / 2열: Y / 3열: R
    #주기가 존재 -> 첫번째 주기만 확인하면됨
    # 노란불이 겹치지 않으면 -1
    
    cycle = 1
    # 1. 각 신호등의 전체 주기를 구한다.
    for G,Y,R in signals:
        temp_cycle = G + Y + R
        cycle = cycle // gcd(cycle,temp_cycle)*temp_cycle
    # 2. 모든 주기의 최소공배수를 구한다.
    # 최소공배수 = a × b ÷ 최대공약수
    # 3. 0부터 전체 반복 주기 직전까지 확인한다.
    for time in range(cycle):
        # 4.현재 시간에 모든 신호등이 노란불인지 검사한다.
        for G,Y,R in signals:
            #현재 신호등의 주기
            temp_cycle = G + Y + R
            #현재 신호등의 주기 안에서 현재 위치
            current_spot = time % temp_cycle
            #현재 시간 중 하나의 신호등이라도 노란색이 아니면 현재 신호등 종료시키고 다음 시간으로 넘어감.
            if not (G < current_spot<=G+Y):
                break
        # 5. 모두 노란불이면 time을 반환한다.
        #내부 반복문이 break 없이 종료되었다면 모든 신호등이 노란불이라는 뜻이므로 time을 반환한다.
        else:
            return time
    # 6. 끝까지 없으면 -1을 반환한다.
    return -1